flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java

```diff
@@ -204,6 +222,11 @@ public static StateBackend fromApplicationOrConfigOrDefault(
 
         // (1) the application defined state backend has precedence
         if (fromApplication != null) {
+
+            checkArgument(
+                    !(fromApplication instanceof DelegateStateBackend),
+                    "DelegateStateBackend can not be delegated!");
```

Comment:
Why not?

Reply:
> Why not?

> Another concern: can we get DelegateStateBackend here after deserialization?

Reply:
When I said why not, I was thinking of why does `deserialization` matter here? Or put it another way, `deserialization` in where for what? And why does `deserialization` be a concern?

If you mean the read config deserialization in JM or TM, it does not matter because we do not allow `changelogstatebackend` from the application. Even if we do, that should not be a problem as well, because DelegateStateBackend is a StateBackend which is `java.io.Serializable`.

Or do you mean something else?

Reply:
>  we do not allow changelogstatebackend from the application.

Do you mean user can NOT use java API, only config? Why?
I think user SHOULD be able to configure backend via both API and config.

I mean a scenario when
1. `ChangelogStateBackend` is created in the user app 
2. then stored (serialized) on JM 
3. then deserialized on TM 
4. the passed to this method

P.S.: this method anyways should not perform the check as it is not aware of **why** this check should be done.
[This](https://github.com/apache/flink/pull/14943#discussion_r583600890) is I think the right place.

Reply:
But does not ALL OF US agree that `ChangelogStateBackend` is only configured through "flag"? That long discussion with Seth, Piotr, Arvid and you??

Please check the COUPLE OF HILIGHTS of this PR
"ChangelogStateBackend is enabled through configuration key "state.backend.enable-statechangelog". It is not directly exposed to users, because ChangelogStateBackend is an implementation detail and users should NOT be aware of it. In other words, application initialization is not supported:
evn.setStateBackend(new ChangelogStateBackend(...)) NOT SUPPORT"

And that's why I added the check 

"""
if (config.get(CheckpointingOptions.ENABLE_STATE_CHANGE_LOG)) {
            LOG.info(
                    "Delegate State Backend is used, and the root State Backend is {}",
                    backend.getClass().getSimpleName());
"""

Even if it allows to configure through the app, I dont think deserialization would be a problem, which is not now. That's why I ask where the concern comes from.

Reply:
We agreed to use "flag" in **both** config and API. I didn't see this API change in this PR and assumed it's addition is postponed and for now explicit configuration will be used in Flink internally.
I think that for internal use explicit configuration will be needed anyways; and to implement the "flag" in the API we'll have to remove this check.

Furthermore, I don't see any reason why should we disallow users (or us) explicit configuration. If someone has already constructed and passed the backend here then it means it's already exposed.

> I dont think deserialization would be a problem. That's why I ask where the concern comes from.

Please see my [reply above](https://github.com/apache/flink/pull/14943#discussion_r583867401).

Reply:
What do you mean by "flag" in API?

Reply:
From the offline discussion you mentioned:
`env.getCheckpointConfig().enableChangelogPersistence();`

> And does the "flag" in config mean the same thing as what is supported in this PR?

I'm not sure if I understand your question. I think this PR doesn't support this  currently; but it should, or at least should allow explicit configuration via API.

Reply:
Let's maybe discuss it offline? I think we have some misunderstandings about what is "flag" means. Or we refer to different things. I guess I kind of understanding, but just in case.